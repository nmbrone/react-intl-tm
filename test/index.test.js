const fs = require('fs');
const path = require('path');
const TranslationsManager = require('../lib');

describe('TranslationsManager', () => {
  const tmpDir = 'test/tmp';
  const translationsDir = path.join(tmpDir, 'translations');

  const messages1 = [
    { id: 'message_1', defaultMessage: 'Message 1' },
    { id: 'message_2', defaultMessage: 'Message 2' },
    { id: 'message_3', defaultMessage: 'Message 3' },
  ];

  const messages2 = [
    { id: 'message_3', defaultMessage: 'Message 3' },
    { id: 'message_4', defaultMessage: 'Message 4' },
  ];

  function translationFilePath(locale) {
    return path.join(translationsDir, locale + '.json');
  }

  function readTranslation(locale) {
    return JSON.parse(fs.readFileSync(translationFilePath(locale), 'utf-8'));
  }

  function writeTranslation(locale, content) {
    fs.writeFileSync(translationFilePath(locale), JSON.stringify(content, null, 2) + '\n');
  }

  beforeAll(() => fs.mkdirSync(tmpDir, { recursive: true }));
  afterAll(() => fs.rmdirSync(tmpDir, { recursive: true }));

  describe('options validation', () => {
    test('throws an error when options are not valid', () => {
      expect(() => new TranslationsManager({})).toThrowErrorMatchingInlineSnapshot(
        `"Please provide one of \\"messages\\", \\"messagesDir\\", or \\"source\\" options"`
      );

      expect(() => new TranslationsManager({ messages: [] })).toThrowErrorMatchingInlineSnapshot(
        `"Please provide \\"translationsDir\\" option"`
      );
    });

    test('reads messages from json files in `messagesDir`', () => {
      const res = new TranslationsManager({
        messagesDir: 'test/fixtures/messages',
        locales: ['en', 'de'],
        defaultLocale: 'en',
        translationsDir,
      }).results();

      expect(res).toHaveLength(2);
      expect(res[0].translation).toMatchSnapshot();
    });
  });

  describe('.writeFiles()', () => {
    test('writes translation files', () => {
      expect(fs.existsSync(translationFilePath('en'))).toBe(false);
      expect(fs.existsSync(translationFilePath('de'))).toBe(false);

      new TranslationsManager({
        messages: messages1,
        locales: ['en', 'de'],
        defaultLocale: 'en',
        translationsDir,
      }).writeFiles();

      expect(readTranslation('en')).toMatchSnapshot();
      expect(readTranslation('de')).toMatchSnapshot();
    });

    test('updates translation files', () => {
      new TranslationsManager({
        messages: messages2,
        locales: ['en', 'de'],
        defaultLocale: 'en',
        translationsDir,
      }).writeFiles();

      expect(readTranslation('en')).toMatchSnapshot();
      expect(readTranslation('de')).toMatchSnapshot();
    });

    test('keeps translations untouched', () => {
      writeTranslation('en', {
        message_3: 'Message three',
        message_4: 'Message four',
      });

      writeTranslation('de', {
        message_3: 'Nachricht drei',
        message_4: 'Nachricht vier',
      });

      new TranslationsManager({
        messages: messages2,
        locales: ['en', 'de'],
        defaultLocale: 'en',
        translationsDir,
      }).writeFiles();

      expect(readTranslation('en')).toMatchSnapshot();
      expect(readTranslation('de')).toMatchSnapshot();
    });
  });

  describe('.report()', () => {
    test('prints the report to the console', () => {
      const log = console.log;
      console.log = jest.fn();

      const tm = new TranslationsManager({
        messagesDir: 'test/fixtures/messages',
        locales: ['en', 'de'],
        defaultLocale: 'en',
        translationsDir,
      });

      tm.report();
      expect(console.log).toHaveBeenCalledTimes(2);
      expect(console.log.mock.calls[0]).toMatchSnapshot();

      tm.report({ short: true });
      expect(console.log.mock.calls[2]).toMatchSnapshot();
      console.log = log;
    });
  });
});
