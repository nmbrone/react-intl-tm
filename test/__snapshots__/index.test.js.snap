// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TranslationsManager .report() prints the report to the console 1`] = `
Array [
  "ðŸ“„ [1m[4m[33mtest/tmp/translations/en.json[39m[24m[22m: (3 added, 2 removed)
(+) Added:
  [32mtitle[39m: [32mApp title[39m
  [32mlink.home[39m: [32mHome[39m
  [32mlink.about[39m: [32mAbout[39m
(-) Removed:
  [31mmessage_3[39m: [31mMessage three[39m
  [31mmessage_4[39m: [31mMessage four[39m
",
]
`;

exports[`TranslationsManager .report() prints the report to the console 2`] = `
Array [
  "ðŸ“„ [1m[4m[33mtest/tmp/translations/en.json[39m[24m[22m: (3 added, 2 removed)",
]
`;

exports[`TranslationsManager .writeFiles() keeps translations untouched 1`] = `
Object {
  "message_3": "Message three",
  "message_4": "Message four",
}
`;

exports[`TranslationsManager .writeFiles() keeps translations untouched 2`] = `
Object {
  "message_3": "Nachricht drei",
  "message_4": "Nachricht vier",
}
`;

exports[`TranslationsManager .writeFiles() updates translation files 1`] = `
Object {
  "message_3": "Message 3",
  "message_4": "Message 4",
}
`;

exports[`TranslationsManager .writeFiles() updates translation files 2`] = `
Object {
  "message_3": "Message 3",
  "message_4": "Message 4",
}
`;

exports[`TranslationsManager .writeFiles() writes translation files 1`] = `
Object {
  "message_1": "Message 1",
  "message_2": "Message 2",
  "message_3": "Message 3",
}
`;

exports[`TranslationsManager .writeFiles() writes translation files 2`] = `
Object {
  "message_1": "Message 1",
  "message_2": "Message 2",
  "message_3": "Message 3",
}
`;

exports[`TranslationsManager options validation reads messages from json files in \`messagesDir\` 1`] = `
Object {
  "link.about": "About",
  "link.home": "Home",
  "title": "App title",
}
`;
